### PR TITLE
IB: 0.2 part III

### LINKED ISSUE #19: 0.2 part III: DML supports `trx_id` instead of   timestamps
### ~~TODO:~~
- ~~check if it is correct to allocate from `table->heap` in `row_insert_for_mysql()`;~~
- ~~test with AUTOINC;~~
- ~~[multi-update](https://github.com/natsys/mariadb_10.2/commit/d42472c8180390015cbc820e9ae1052202a56e65)~~;
- ~~test all 'is_max()' cases;~~
- ~~is it correct to use UPD_NODE_NO_ORD_CHANGE in UPDATE?~~
  - Yes, because it's calculated on lower level (`row_upd()`, requires `in_mysql_interface`).

### PR DESCRIPTION
**Pull request by midenok**
- versioned DML: INSERT, UPDATE, DELETE;
- general refactoring and fixes.

Warning: breaks tests `insert` and `update` tests since they require _part IV_.

Closes #19

### CODE REVIEW

**Comment by midenok** [comment]
Please, review.

File: sql/table.cc [hunk 1]
```diff
@@ -2125,19 +2096,70 @@ int TABLE_SHARE::init_from_binary_frm_image(THD *thd, bool write,
...
+      if (vers_end_field()->type() != MYSQL_TYPE_TIMESTAMP)
+      {
+        my_error(ER_SYS_END_FIELD_MUST_BE_TIMESTAMP, MYF(0), share->table_name);
+        goto err;
+      }
+    } // if (db_type()->versioned())
+  } // if (system_period == NULL)
+
+  delete handler_file;
+
```
Comment (avodaniel) [inline sql/table.cc:2155]: Why was it necessary to move the code?

**Comment by avodaniel** [comment]
I like the changes I think you can go ahead.

File: sql/table.cc (hunk 1 above)
Comment (midenok) [reply sql/table.cc:2155]: `db_type()` requires `share->db_plugin`.

**Comment by midenok** [comment]
@kevgs it's not too late to leave comments here in case you will have any.

File: sql/field.cc [hunk 2]
```diff
@@ -4355,6 +4355,20 @@ void Field_longlong::sql_type(String &res) const
   add_zerofill_and_unsigned(res);
 }
 
+bool Field_longlong::set_max()
+{
+  ASSERT_COLUMN_MARKED_FOR_WRITE_OR_COMPUTED;
+  int8store(ptr, ULONGLONG_MAX);
```
Comment (kevgs) [inline sql/field.cc:4361]: Why mixing signed and unsigned here and in Field_longlong::is_max()?

Comment (midenok) [reply sql/field.cc:4361]: Where exactly do you see mixing signed and unsigned?

Comment (kevgs) [reply sql/field.cc:4361]: Storing ULONGLONG_MAX for signed Field_longlong class and reading signed in unsigned variable here
`ulonglong j;`
`j = sint8korr(ptr);`

looks strange.

Comment (midenok) [inline sql/field.cc]: Fixed: c27c6bb915a047a70c1f3c2ec129f9fd4b58ce5d

Comment (midenok) [reply sql/field.cc:4361]: Fixed: 3647e2efae39f8bb8ed8566742a7f36d626362f5

**Pull request merged**